### DISCUSSION TITLE
Uninstaller sanitation

### DISCUSSION THREAD

**Discussion by pljones** [jamulussoftware/jamulus#2724] [Ideas]
Following a question from @gilgongo and subsequent chat...

It seems that maybe the Jamulus uninstaller isn't as clean as it could be in removing things.  However, there could, perhaps, be other improvements too.

The goal of the uninstaller is to remove all traces of Jamulus that _it_ installed.

If there are things in folders it would normally remove that the installer didn't put there, it should report them and offer to remove them.  (There _really_ should be nothing in the Jamulus %PROGRAMFILES% folder the installer didn't put there, so the user should be gently encouraged to investigate and clean up...)

There are three main areas: %PROGRAMFILES% for the main executable and libraries, %APPDATA% for inifiles and the registry, mostly for installer information - but there could be other stuff there I don't know about.

Uninstalling should aim always to remove the version of Jamulus in %PROGRAMFILES%.

Optionally, any %APPDATA% can be removed.  (The user should get a clear warning about this, indicating it's not necessary and they'd be able to reuse settings if they did reinstall, even if it was another version.)

Optionally, the uninstaller can also strip the registry of any record of _this_ install of Jamulus.

If the registry information isn't removed, then the user choices for uninstalling can be retained for future use.

**Comment by ann0see** [comment]
Actually I think the uninstaller used to (and should still) remove registry entries, all data the installer wrote into Program Files but not necessarily the .ini files. However it does not remove files from pre 3.6.2 installs (as far as I remember that was the first release with the new windows installer).

**Reply by gilgongo** [reply]
I get the following after uninstalling 3.9.0 beta1 and installing the previous release:

![image](https://user-images.githubusercontent.com/4561747/178150925-05fe02f6-830b-421a-9c4c-da8ee3b3c970.png)

**Reply by pljones** [reply]
I'm not sure if the addition of Jack for Windows broke it - but I think I vaguely remember something around that change having an effect.

**Reply by ann0see** [reply]
@henkdegroot do you still remember anything around this topic?

**Reply by ann0see** [reply]
@gilgongo what was in this folder?

**Reply by henkdegroot** [reply]
I will need to check,, never had this issue on my systems (and I regular use the installer for install/unnstall). Maybe happens when custom files are stored in the program files folder?

**Reply by ann0see** [reply]
Yes. Custom files aren’t removed

**Reply by henkdegroot** [reply]
Actually, I don't think I have unstalled a 390beta yet. So will check what happens on my system doing that.

EDIT: Tested and uninstall is removing the Jamulus folder on my system.

**Reply by pljones** [reply]
OK...
1. Did previous versions complain about the directory already existing?  If not, this is not a blocker for 3.9.0.
2. Should the installer refuse to install into an existing directory?  Maybe it should just ask for confirmation?  (Writing "I really know what I'm doing" into a text field ;) ).  If so, this would be an change for a later release.

So only if something has changed should this be worried about now.

**Reply by henkdegroot** [reply]
The behaviour of the previous installers is the same. When the folder exists, you need to use a different one.
It would be good improvement to allow the installation in the same folder in a future release.
